### ISSUE TITLE
Exception classes

### ISSUE THREAD

**Issue by mkulagowski** [mkulagowski/MineZPRft#10]
- [x] Add main Exception class that inherits from std::exception.
- [ ] Declare exceptions for all used classes

**Comment by lookeypl** [comment]
I was thinking about using different exception objects for different failures, something like in Java.

Examples:
- Shaders fail to load from disk. Instead of throwing general RendererException we throw ShaderFileException.
- Renderer fails to initialize OpenGL Extensions - throw OGLExtensionException
- Terrain Generator cannot load the map - throw TerrainLoadException
- etc etc etc

Naturally the class should keep information about what is happening, an error message in log of some sorts, which will be added when throwing. What do you think about it?

**Comment by mkulagowski** [comment]
Sounds like a lot of virtual and inheritance - would be great in this project to have some practical use ;d

So the main idea is:
1. We catch Exception object
2. We do some `LOG_ERROR("place, function, msg: %s", except_obj->ErrorMsg());`
3. We hope that `ErrorMsg()` will be different for each exception subclass

Still I'd do sth like :

```
       Exception
     /      |     \
Renderer Common TerrainMgr ...
         /    \ 
       Math  FileSystem
        /         \
DivisionByZero    WrongPath
```

So that calling `ShaderFileException->ErrorMsg()` we get msg like "Exception/Renderer/Shader - Shaders failed to load from disk", where "Exception" is from Exception and "/Renderer" is from Renderer : Exception

**Comment by mkulagowski** [comment]
Main exception [commit](https://review.gerrithub.io/#/c/254129)

**Comment by mkulagowski** [comment]
Exception class is now implemented. Now we've to use it. Let's start by reworking current code, then just use exceptions in any future updates ;)
How about it?

About usage:
When we use exceptions in e.x. Window class, then we use DECLARE_EXCEPTION macro in Window.hpp file: 
- under any existing macros
- over class declaration
- add `// Used exceptions` comment above.

**Comment by lookeypl** [comment]
Agreed. I will go through Renderer and try to exceptionize it as much as possible.

Moreover, we must have some kind of try...catch in the code. For now I suggest adding it in main.cpp, it should catch std::exceptions and give us an information what went wrong.

**Comment by mkulagowski** [comment]
Wouldn't adding a try...catch only in main.cpp result in any throw practically quitting the main loop?

I think we can add try..catch in main.cpp, BUT only for errors that force us to terminate the application. Other try..catches would get inside the main loop and just throw LOG_ERROR or LOG_WARNING, yet still continue working and only rethrow if error is one of fatal ones - we can declare "FatalException" and declare all fatal exceptions as its subexceptions ;)
E.x. We can't connect with Xserver for window creation - terminating app is a must (with a log_error ofc ;) ). But we can't load one of the textures? We can still continue and operate, but error will be logged.

How about that? (not sure about example, because all ogl management is more of Your thing :P)

**Comment by lookeypl** [comment]
As I wrote: "For now". Please, read with caution.

Further try...catches will be added, however we need some kind of security for now to avoid termination due to exception flying outside of main.

**Comment by lookeypl** [comment]
And, as a sidenote - there is a reason why all main loop functions added by me have "noexcept" suffix. ;)
